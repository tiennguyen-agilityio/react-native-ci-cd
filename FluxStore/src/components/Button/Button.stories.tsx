import {Meta, StoryObj} from '@storybook/react';

import Button from '.';
import {Alert} from 'react-native';

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
};

export default meta;

type Story = StoryObj<typeof Button>;

export const Default: Story = {
  args: {
    text: 'Click me!',
    onPress: () => Alert.alert('Notifications'),
  },
};

export const Outlined: Story = {
  args: {
    text: 'Click me!',
    variant: 'outlined',
    onPress: () => Alert.alert('Notifications'),
  },
};

export const Ghost: Story = {
  args: {
    text: 'Click me!',
    variant: 'ghost',
    onPress: () => Alert.alert('Notifications'),
  },
};

export const small: Story = {
  args: {
    text: 'Click me!',
    size: 'sm',
    onPress: () => Alert.alert('Notifications'),
  },
};
