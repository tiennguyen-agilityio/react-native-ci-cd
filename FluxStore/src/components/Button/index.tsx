import {memo, ReactNode, useMemo} from 'react';
import {
  TouchableOpacity,
  Text,
  PressableProps,
  ColorValue,
  DimensionValue,
  ViewStyle,
  StyleSheet,
  TextStyle,
  ActivityIndicator,
} from 'react-native';

// Stores
import {useThemeStore} from '@/stores';

// Themes
import {colors, fontSizes, fontWeights, fontFamilies} from '@/themes';

export const buttonSize = StyleSheet.create({
  sm: {
    height: 48,
  },

  md: {
    height: 51,
  },

  lg: {
    height: 53,
  },
});

export const buttonTextSize = StyleSheet.create({
  sm: {
    fontSize: fontSizes.sm,
  },

  md: {
    fontSize: fontSizes.sm,
  },

  lg: {
    fontSize: fontSizes.base,
  },
});

export const buttonVariants = StyleSheet.create({
  solid: {
    backgroundColor: colors.black[500],
  },

  outlined: {
    backgroundColor: colors.gray[500],
    borderWidth: 1,
    borderColor: colors.white[500],
  },

  ghost: {
    backgroundColor: colors.transparent,
  },
});

export const buttonTextVariants = StyleSheet.create({
  solid: {
    color: colors.white[500],
  },

  outlined: {
    color: colors.white[500],
  },

  ghost: {
    color: colors.green[200],
    fontWeight: fontWeights.regular,
  },
});

export const buttonBasicStyles: ViewStyle = {
  flexDirection: 'row',
  justifyContent: 'center',
  alignContent: 'center',
  alignItems: 'center',
  gap: 15,
  borderRadius: 25,
};

export const buttonTextBasicStyles: TextStyle = {
  textAlign: 'center',
  borderRadius: 25,
  fontFamily: fontFamilies.productSans.bold,
  fontWeight: fontWeights.bold,
};

export interface ButtonProps extends PressableProps {
  text: string;
  disabled?: boolean;
  isLoading?: boolean;
  variant?: 'solid' | 'outlined' | 'ghost';
  size?: 'sm' | 'md' | 'lg';
  color?: ColorValue;
  width?: DimensionValue;
  height?: DimensionValue;
  fontSize?: number;
  textDecorationLine?: string;
  startIcon?: ReactNode;
  style?: ViewStyle;
  onPress?: () => void;
}

const Button = ({
  text,
  variant = 'solid',
  size = 'md',
  width = variant === 'ghost' ? 'auto' : '100%',
  height,
  disabled,
  color,
  fontSize,
  startIcon,
  style,
  textDecorationLine = 'none',
  isLoading = false,
  onPress,
}: ButtonProps) => {
  const {theme, isDark} = useThemeStore();

  const bgButton = useMemo(() => {
    if (disabled && variant !== 'ghost')
      return {
        backgroundColor: theme.background.secondary,
      };

    return variant === 'solid' && isDark
      ? {
          backgroundColor: theme?.primary,
        }
      : null;
  }, [disabled, variant, theme, isDark]);

  const textColor = useMemo(() => {
    return variant === 'solid' && isDark
      ? {
          color: theme?.text?.light,
        }
      : null;
  }, [variant, theme, isDark]);

  return (
    <TouchableOpacity
      style={[
        buttonBasicStyles,
        buttonVariants[variant],
        buttonSize[size],
        {
          ...(width && {width}),
          ...(height && {height}),
          ...(bgButton && bgButton),
        },
        style,
      ]}
      activeOpacity={0.8}
      disabled={disabled}
      onPress={onPress}
      testID="button">
      {startIcon && startIcon}
      <Text
        style={[
          buttonTextBasicStyles,
          buttonTextVariants[variant],
          buttonTextSize[size],
          {
            ...(textColor && textColor),
            ...(color && {color}),
            ...(fontSize && {fontSize}),
            ...(textDecorationLine && {textDecorationLine}),
          } as TextStyle,
        ]}>
        {text}
      </Text>
      {isLoading && <ActivityIndicator size="small" color={theme.secondary} style={{}} />}
    </TouchableOpacity>
  );
};

export default memo(Button);
