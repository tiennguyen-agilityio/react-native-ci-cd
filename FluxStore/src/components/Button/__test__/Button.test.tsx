import {render, fireEvent, screen} from '@testing-library/react-native';

import Button from '..';

const onPress = jest.fn();

describe('Button', () => {
  const props = {
    text: 'Click me',
    onPress,
  };

  it('should render button component with default props', () => {
    const {toJSON} = render(<Button disabled {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render button component with disabled', () => {
    const {toJSON} = render(<Button disabled {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onClick when being clicked', () => {
    render(<Button testID="button" {...props} />);

    const button = screen.getByTestId('button');
    fireEvent.press(button);

    expect(onPress).toHaveBeenCalled();
  });
});
