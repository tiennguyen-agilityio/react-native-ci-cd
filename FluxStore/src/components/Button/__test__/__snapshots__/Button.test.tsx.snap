// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button component with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignContent": "center",
      "alignItems": "center",
      "backgroundColor": "#9d9d9d",
      "borderRadius": 25,
      "flexDirection": "row",
      "gap": 15,
      "height": 51,
      "justifyContent": "center",
      "opacity": 1,
      "width": "100%",
    }
  }
  testID="button"
>
  <Text
    style={
      [
        {
          "borderRadius": 25,
          "fontFamily": "ProductSansBold",
          "fontWeight": 700,
          "textAlign": "center",
        },
        {
          "color": "#fff",
        },
        {
          "fontSize": 16,
        },
        {
          "textDecorationLine": "none",
        },
      ]
    }
  >
    Click me
  </Text>
</View>
`;

exports[`Button should render button component with disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignContent": "center",
      "alignItems": "center",
      "backgroundColor": "#9d9d9d",
      "borderRadius": 25,
      "flexDirection": "row",
      "gap": 15,
      "height": 51,
      "justifyContent": "center",
      "opacity": 1,
      "width": "100%",
    }
  }
  testID="button"
>
  <Text
    style={
      [
        {
          "borderRadius": 25,
          "fontFamily": "ProductSansBold",
          "fontWeight": 700,
          "textAlign": "center",
        },
        {
          "color": "#fff",
        },
        {
          "fontSize": 16,
        },
        {
          "textDecorationLine": "none",
        },
      ]
    }
  >
    Click me
  </Text>
</View>
`;
